File: Web/Infrastructure/Exceptions/ApiErrorResponse.cs
namespace Web.Exceptions;

public class ApiErrorResponse
{
    public string? Message { get; set; }
    public string? StackTrace { get; set; }
    public string? UserVisibleMessage { get; set; }

    public ApiErrorResponse(string? message, string? stackTrace, string? userVisibleMessage)
    {
        Message = message;
        StackTrace = stackTrace;
        UserVisibleMessage = userVisibleMessage;
    }
}

File: ProjectName/Web/client-app/src/api/ApiErrorResponse.ts
export type ApiErrorResponse = {
  message: string | null;
  stackTrace: number | null;
  userVisibleMessage: string | null;
};

const emptyApiErrorResponse: ApiErrorResponse = {
  message: null,
  stackTrace: null,
  userVisibleMessage: null,
};
export const isApiErrorResponse = (
  value: unknown,
): value is ApiErrorResponse => {
  if (typeof value !== "object") {
    return false;
  }

  for (const propertyName of Object.getOwnPropertyNames(
    emptyApiErrorResponse,
  )) {
    if (!Object.prototype.hasOwnProperty.call(value, propertyName)) {
      return false;
    }
  }

  return true;
};
